pkg/controller/instance/controller_reconcile.go

```diff
@@ -123,193 +155,117 @@ func (igr *instanceGraphReconciler) reconcileInstance(ctx context.Context) error
-	if want, err := igr.runtime.ReadyToProcessResource(resourceID); err != nil || !want {
-		log.V(1).Info("Skipping resource processing", "reason", err)
-		resourceState.State = ResourceStateSkipped
-		igr.runtime.IgnoreResource(resourceID)
-		return nil
-	}
-
-	// Get and validate resource state
-	resource, state := igr.runtime.GetResource(resourceID)
-	if state != runtime.ResourceStateResolved {
-		return igr.delayedRequeue(fmt.Errorf("resource %s not resolved: state=%v", resourceID, state))
-	}
+	unresolvedResourceID := ""
+	prune := true
+	// Reconcile resources in topological order
+	for _, resourceID := range igr.runtime.TopologicalOrder() {
+		log := igr.log.WithValues("resourceID", resourceID)
+
+		// Initialize resource state in instance state
+		resourceState := &ResourceState{State: ResourceStateInProgress}
+		igr.state.ResourceStates[resourceID] = resourceState
+
+		// Check if resource should be processed (create or get)
+		// TODO(barney-s): skipping on error seems un-intuitive, should we skip on CEL evaluation error?
+		if want, err := igr.runtime.ReadyToProcessResource(resourceID); err != nil || !want {
+			log.V(1).Info("Skipping resource processing", "reason", err)
+			resourceState.State = ResourceStateSkipped
+			igr.runtime.IgnoreResource(resourceID)
+			continue
+		}
 
-	// Handle resource reconciliation
-	return igr.handleResourceReconciliation(ctx, resourceID, resource, resourceState)
-}
+		// Check if the resource dependencies are resolved and can be reconciled
+		resource, state := igr.runtime.GetResource(resourceID)
 
-// handleResourceReconciliation manages the reconciliation of a specific resource,
-// including creation, updates, and readiness checks.
-func (igr *instanceGraphReconciler) handleResourceReconciliation(
-	ctx context.Context,
-	resourceID string,
-	resource *unstructured.Unstructured,
-	resourceState *ResourceState,
-) error {
-	log := igr.log.WithValues("resourceID", resourceID)
+		if state != runtime.ResourceStateResolved {
+			unresolvedResourceID = resourceID
+			prune = false
+			break
+		}
 
-	// Get resource client and namespace
-	rc := igr.getResourceClient(resourceID)
+		applyable := applyset.ApplyableObject{
+			Unstructured: resource,
+			ID:           resourceID,
+			ExternalRef:  igr.runtime.ResourceDescriptor(resourceID).IsExternalRef(),
+		}
+		clusterObj, err := aset.Add(ctx, applyable)
+		if err != nil {
+			return fmt.Errorf("failed to add resource to applyset: %w", err)
+		}
 
-	// Check if resource exists
-	observed, err := rc.Get(ctx, resource.GetName(), metav1.GetOptions{})
-	if err != nil {
-		if apierrors.IsNotFound(err) {
-			// For read-only resources, we don't create
-			if igr.runtime.ResourceDescriptor(resourceID).IsExternalRef() {
-				resourceState.State = "WAITING_FOR_EXTERNAL_RESOURCE"
-				resourceState.Err = fmt.Errorf("external resource not found: %w", err)
-				return igr.delayedRequeue(resourceState.Err)
+		if clusterObj != nil {
```

Comment:
will cluster can't be nil here.

Reply:
it can be nil if it is applied for the first time. Because we havent yet applied. We just added to applyset. apply happens in 218.